external/fv3fit/fv3fit/pytorch/cyclegan/cyclegan_trainer.py

```diff
@@ -0,0 +1,424 @@
+            itertools.chain(
+                generator_a_to_b.parameters(), generator_b_to_a.parameters()
+            )
+        )
+        optimizer_discriminator = self.discriminator_optimizer.instance(
+            itertools.chain(discriminator_a.parameters(), discriminator_b.parameters())
+        )
+        return CycleGANTrainer(
+            cycle_gan=CycleGAN(
+                model=CycleGANModule(
+                    generator_a_to_b=generator_a_to_b,
+                    generator_b_to_a=generator_b_to_a,
+                    discriminator_a=discriminator_a,
+                    discriminator_b=discriminator_b,
+                ).to(DEVICE),
+                state_variables=state_variables,
+                scalers=_merge_scaler_mappings(scalers),
+            ),
+            optimizer_generator=optimizer_generator,
+            optimizer_discriminator=optimizer_discriminator,
+            identity_loss=self.identity_loss.instance,
+            cycle_loss=self.cycle_loss.instance,
+            gan_loss=self.gan_loss.instance,
+            batch_size=n_batch,
+            identity_weight=self.identity_weight,
+            cycle_weight=self.cycle_weight,
+            generator_weight=self.generator_weight,
+            discriminator_weight=self.discriminator_weight,
+        )
+
+
+def _merge_scaler_mappings(
+    scaler_tuple: Tuple[Mapping[str, StandardScaler], Mapping[str, StandardScaler]]
+) -> Mapping[str, StandardScaler]:
+    scalers = {}
+    for prefix, scaler_map in zip(("a_", "b_"), scaler_tuple):
+        for key, scaler in scaler_map.items():
+            scalers[prefix + key] = scaler
+    return scalers
+
+
+class ReplayBuffer:
+
+    # To reduce model oscillation during training, we update the discriminator
+    # using a history of generated data instead of the most recently generated data
+    # according to Shrivastava et al. (2017).
+
+    def __init__(self, max_size=50):
+        if max_size <= 0:
+            raise ValueError("max_size must be positive")
+        self.max_size = max_size
+        self.data = []
+
+    def push_and_pop(self, data: torch.Tensor) -> torch.autograd.Variable:
+        """
+        Push data into the buffer and return a random sample of the buffer.
+
+        If there are at least max_size elements in the buffer, the returned sample
+        is removed from the buffer.
+        """
+        to_return = []
+        for element in data.data:
+            element = torch.unsqueeze(element, 0)
+            if len(self.data) < self.max_size:
+                self.data.append(element)
+                to_return.append(element)
+            else:
+                if random.uniform(0, 1) > 0.5:
+                    i = random.randint(0, self.max_size - 1)
+                    to_return.append(self.data[i].clone())
+                    self.data[i] = element
+                else:
+                    to_return.append(element)
+        return torch.autograd.Variable(torch.cat(to_return))
```

Comment:
Is this the right API to be using? https://pytorch.org/docs/stable/autograd.html#variable-deprecated

Reply:
Will be swapping this code out.

Reply:
Swapped.